pkg/hope/node_management.go

```diff
@@ -101,9 +102,90 @@ func CreateClusterMaster(log *logrus.Entry, node *Node, podNetworkCidr string, f
 		}
 	}
 
+	// Update the load balancer before even starting the process.
+	// Trying to time it with the init process will be clumsy, so yeet.
+	if loadBalancer != nil {
+		SetLoadBalancerHosts(log, loadBalancer, masters)
+	}
+
+	// Search through the existing masters to see if this node is being added
+	//   as a master to an existing control plane, or if this will be the
+	//   first master in the pool.
 	connectionString := node.ConnectionString()
+	for _, aMaster := range *masters {
+		aMasterCs := aMaster.ConnectionString()
+		if aMasterCs == connectionString {
+			continue
+		}
+
+		certKey := ""
+
+		if loadBalancer != nil {
+			remoteAdminConfPath := "/etc/kubernetes/admin.conf"
+			grepRegexp := "'\\s+server: https://api\\.internal\\.aleemhaji\\.com:6443'"
+
+			if err := ssh.ExecSSH(aMasterCs, "grep", "-E", grepRegexp, "-q", remoteAdminConfPath); err != nil {
+				log.Warn("Other master node", aMaster.Host, "isn't connected to load balancer.")
+				continue
+			}
+
+			// From this master node, pull a control plane certificate key, and
+			//   use it to run the kubeadm join command.
+			output, err := ssh.GetSSH(aMasterCs, "kubeadm", "init", "phase", "upload-certs", "--upload-certs")
+			if err != nil {
+				return err
+			}
+
+			for _, line := range strings.Split(output, "\n") {
+				line = strings.TrimSpace(line)
+				match, err := regexp.MatchString("[0-9a-f]{64}", line)
+				if err != nil {
+					return err
+				}
+
+				if match {
+					certKey = line
+					break
+				}
+			}
+
+			if certKey == "" {
+				return fmt.Errorf("Failed to find cert key from selected master node: %s", aMaster)
+			}
+		}
+
+		joinCommand, err := ssh.GetSSH(aMasterCs, "kubeadm", "token", "create", "--print-join-command")
+		if err != nil {
+			return err
+		}
+
+		joinComponents := strings.Split(strings.TrimSpace(joinCommand), " ")
+		allArguments := append([]string{node.ConnectionString()}, joinComponents...)
+
+		if loadBalancer != nil {
+			allArguments = append(allArguments, "--control-plane", "--certificate-key", certKey)
+		}
+
+		if err := ssh.ExecSSH(allArguments...); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// At this point, it's very probable that the node being added is the
+	//   first master node to be created in the cluster.
+	// Note: While this is running, the load balancer has to be updated to
+	//   include the master host.
+	// Maybe wait 20 seconds before updating the lb.
 	podNetworkCidrArg := fmt.Sprintf("--pod-network-cidr=%s", podNetworkCidr)
-	if err := ssh.ExecSSH(connectionString, "sudo", "kubeadm", "init", podNetworkCidrArg); err != nil {
+	allArgs := []string{node.ConnectionString(), "sudo", "kubeadm", "init", podNetworkCidrArg}
+	if loadBalancer != nil {
+		loadBalancerEndpoint := fmt.Sprintf("%s:%s", loadBalancerHost, "6443")
```

Comment:
Should be constructed way sooner and used in place of the existing hard coded piece.